File: retiro-del-rocio/app/Models/Asset.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Carbon;

/**
 * A registered hotel asset (equipment/fixture) maintenance tracks service
 * history and a preventive-maintenance schedule for — e.g. a room's AC unit,
 * the lobby generator, a lift.
 */
class Asset extends Model
{
    protected $fillable = [
        'name', 'category', 'room_unit_id', 'location_label', 'notes',
        'service_interval_days', 'last_serviced_at',
    ];

    protected $casts = [
        'last_serviced_at' => 'datetime',
        'service_interval_days' => 'integer',
    ];

    public function roomUnit()
    {
        return $this->belongsTo(RoomUnit::class);
    }

    public function workOrders()
    {
        return $this->hasMany(WorkOrder::class);
    }

    /** Where this asset lives — its room, or its free-text location. */
    public function locationLabel(): string
    {
        $unit = $this->relationLoaded('roomUnit') ? $this->roomUnit : $this->roomUnit()->first();

        if ($unit?->number) {
            return 'Room '.$unit->number;
        }

        return $this->location_label ?: 'Hotel-wide';
    }

    /** No schedule tracked, or the interval hasn't elapsed since the last service (or creation, if never serviced). */
    public function isDueForService(): bool
    {
        if (! $this->service_interval_days) {
            return false;
        }

        $since = $this->last_serviced_at ?? $this->created_at;

        return $since !== null && $since->addDays($this->service_interval_days)->isPast();
    }

    /** null when not on a PM schedule. */
    public function nextServiceDueAt(): ?Carbon
    {
        if (! $this->service_interval_days) {
            return null;
        }

        $since = $this->last_serviced_at ?? $this->created_at;

        return $since?->copy()->addDays($this->service_interval_days);
    }

    public function toMaintenanceArray(): array
    {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'category' => $this->category,
            'location_label' => $this->locationLabel(),
            'notes' => $this->notes,
            'service_interval_days' => $this->service_interval_days,
            'last_serviced_at' => $this->last_serviced_at?->toIso8601String(),
            'last_serviced_label' => $this->last_serviced_at?->diffForHumans(),
            'next_service_due_at' => $this->nextServiceDueAt()?->toIso8601String(),
            'is_due_for_service' => $this->isDueForService(),
        ];
    }

    /** The Assets screen's service-history list for this asset — every work order ever raised against it, newest first. */
    public function toMaintenanceDetailArray(): array
    {
        $history = $this->relationLoaded('workOrders')
            ? $this->workOrders
            : $this->workOrders()->latest('id')->limit(50)->get();

        return [
            ...$this->toMaintenanceArray(),
            'service_history' => $history->sortByDesc('id')->values()->map->toMaintenanceArray()->values(),
        ];
    }
}

File: retirodelrocioapp/lib/features/maintenance/domain/asset.dart
import 'package:flutter/foundation.dart';
import 'package:retirodelrocioapp/features/maintenance/domain/work_order.dart';

/// A registered hotel asset — its profile, service-history summary, and
/// preventive-maintenance due state (`GET /maintenance/assets`).
@immutable
class Asset {
  const Asset({
    required this.id,
    required this.name,
    this.category,
    required this.locationLabel,
    this.notes,
    this.serviceIntervalDays,
    this.lastServicedAt,
    this.lastServicedLabel,
    this.nextServiceDueAt,
    required this.isDueForService,
    this.serviceHistory = const [],
  });

  final int id;
  final String name;
  final String? category;
  final String locationLabel;
  final String? notes;
  final int? serviceIntervalDays;
  final DateTime? lastServicedAt;
  final String? lastServicedLabel;
  final DateTime? nextServiceDueAt;
  final bool isDueForService;

  /// Populated only on the asset detail response.
  final List<WorkOrder> serviceHistory;

  bool get isOnScheduledMaintenance => serviceIntervalDays != null;

  factory Asset.fromJson(Map<String, dynamic> json) => Asset(
    id: (json['id'] as num?)?.toInt() ?? 0,
    name: json['name'] as String? ?? 'Asset',
    category: json['category'] as String?,
    locationLabel: json['location_label'] as String? ?? 'Hotel-wide',
    notes: json['notes'] as String?,
    serviceIntervalDays: (json['service_interval_days'] as num?)?.toInt(),
    lastServicedAt: json['last_serviced_at'] != null
        ? DateTime.tryParse(json['last_serviced_at'] as String)
        : null,
    lastServicedLabel: json['last_serviced_label'] as String?,
    nextServiceDueAt: json['next_service_due_at'] != null
        ? DateTime.tryParse(json['next_service_due_at'] as String)
        : null,
    isDueForService: json['is_due_for_service'] as bool? ?? false,
    serviceHistory: ((json['service_history'] as List?) ?? const [])
        .map((o) => WorkOrder.fromJson((o as Map).cast()))
        .toList(),
  );
}
